### PR TITLE
[CLIENT-4704] Fix memory leak where adding an operation fails because of the bin name being too long

### PR DESCRIPTION
**Pull request by juliannguyen4**
## Extra Changes
- CI/CD: add smoke tests

## Notes

- Everywhere `as_cdt_add_packed` is called, a stack allocated `as_packer pk` is initialized with all zeroes and passed to that function
- New `list_check_bin_name_length_handling` test case passes
- In the Python client's CI/CD pipeline, valgrind no longer shows the memory leak

### CODE REVIEW

File: src/main/aerospike/as_cdt_internal.c [hunk 1]
```diff
@@ -14,15 +14,53 @@
...
+
 as_binop*
 as_binop_forappend(as_operations* ops, as_operator operator, const char* name);
 
+
+// --------------------------------------------------------------------------------
+// Local helpers.
+
+/**
+ * @private Remove buffer attached to a packer.
```
Comment (juliannguyen4) [inline src/main/aerospike/as_cdt_internal.c]:
```suggestion
 * @private Remove buffer attached to an as_packer variable.
```

File: src/main/aerospike/as_cdt_internal.c [hunk 2]
```diff
@@ -14,15 +14,53 @@
...
+	if (pk->buffer) {
+		cf_free(pk->buffer);
+		pk->buffer = NULL;
+	}
+}
+
+
+// --------------------------------------------------------------------------------
+// Public API
+
```
Comment (juliannguyen4) [inline src/main/aerospike/as_cdt_internal.c]:
```suggestion
```
I don't think these should be labelled as public API methods. They're not marked with `AS_EXTERN`

File: src/test/aerospike_list/list_basics.c [hunk 3]
```diff
@@ -3938,6 +3938,82 @@ TEST(list_select_tree, "test select tree")
...
 
+TEST(list_check_bin_name_length_handling, "test bin name length handling")
+{
+	// This test aims to reproduce an edge case found during Python testing,
+	// where as_cdt_end() causes memory to be allocated but not freed as a
+	// result of an error from as_cdt_add_packed().  See CLIENT-4704.
+
+	// First, compute a bin name that is too long to process.
+	char long_bin_name[128];
+	memset(long_bin_name, 0, sizeof(long_bin_name));
```
Comment (juliannguyen4) [inline src/test/aerospike_list/list_basics.c:3949]:
```suggestion
```
nit: maybe this step is unnecessary if `sprintf` already sets `long_bin_name` to a null-terminated string

File: src/test/aerospike_list/list_basics.c [hunk 4]
```diff
@@ -3938,6 +3938,82 @@ TEST(list_select_tree, "test select tree")
...
+	as_bin* bin = &prec->bins.entries[0];
+
+	assert_int_eq(as_bin_get_type(bin), AS_LIST);
+	as_list* l = (as_list*)as_bin_get_value(bin);
+	assert_int_eq(as_list_size(l), 5);
+	assert_int_eq(as_list_get_int64(l, 0), 40);
+	assert_int_eq(as_list_get_int64(l, 1), 6);
+	assert_int_eq(as_list_get_int64(l, 2), 13);
+	assert_int_eq(as_list_get_int64(l, 3), 27);
+	assert_int_eq(as_list_get_int64(l, 4), 33);
```
Comment (juliannguyen4) [inline src/test/aerospike_list/list_basics.c]: The list elements can be stored in an array and reused for both the earlier put() and or this check

File: src/test/aerospike_list/list_basics.c (hunk 3 above)
Comment (sfalvo-aerospike) [reply src/test/aerospike_list/list_basics.c:3949]: I think it's technically true, but I am playing it safe for safety's sake.

**Review by BrianNichols — APPROVED** [review]
